test/jdk/java/net/httpclient/offline/FixedResponseHttpClient.java

```diff
@@ -240,4 +259,54 @@ public void onSubscribe(Flow.Subscription subscription) {
 
         @Override public void onComplete() { /* do nothing */ }
     }
+
+    @Override
+    public boolean isTerminated() {
+        synchronized (this) {
+            if (!shutdownRequested) return false;
+            return responses.isEmpty();
+        }
+    }
+
+    @Override
+    public void shutdown() {
+        shutdownRequested = true;
+    }
+
+    @Override
+    public void shutdownNow() {
+        shutdown();
+        CompletableFuture[] futures;
+        synchronized (this) {
+            if (responses.isEmpty()) return ;
+            futures = responses.toArray(CompletableFuture[]::new);
+            responses.clear();
+        }
+        for (var op : futures) {
+            op.cancel(true);
+        }
+    }
+
+    @Override
+    public boolean awaitTermination(Duration duration) throws InterruptedException {
+        Objects.requireNonNull(duration);
+        CompletableFuture[] futures;
+        synchronized (this) {
+            if (responses.isEmpty()) return true;
+            futures = responses.toArray(CompletableFuture[]::new);
+        }
+        var all = CompletableFuture.allOf(futures);
+        Duration max = Duration.ofMillis(Long.MAX_VALUE);
+        long timeout = duration.compareTo(max) > 0 ? Long.MAX_VALUE : duration.toMillis();
+        try {
+            all.exceptionally((t) -> null).get(timeout, TimeUnit.MILLISECONDS);
```

Comment:
This implementation of `awaitTermination` only waits for the current (snapshot) of requests that are currently in progress. If a new request happens to be initiated (and is in-progress) after this snapshot of `responses.toArray()` was taken, then this implementation of `awaitTermination` can possibly return an incorrect `true` (indicating the client has terminated) while there currently might be some requests in progress. This `FixedResponseHttpClient` is just a test specific implementation which appears to be used in one single test class (`OfflineTesting`) and as far as I can see, that test won't be impacted with the implementation of `awaitTermination()`, so I think this implementation might be OK for now.

Reply:
Good point.

Reply:
OK - this ended up to be a bit more complicated than I thought. What I had missed is that the `FixedResponseHttpClient` actually wraps another (real) `HttpClient`, so close etc... need to close both. I have now fixed this and updated the offline test to close its clients too.